torch/csrc/cuda/comm.cpp

```diff
@@ -0,0 +1,99 @@
+#include "comm.h"
+
+#include "torch/csrc/utils/tensor_flatten.h"
+#include "torch/csrc/utils/auto_gpu.h"
+#include "torch/csrc/cuda/device_set.h"
+#include "torch/csrc/cuda/nccl.h"
+
+#include <ATen/ATen.h>
+
+namespace torch { namespace cuda {
+
+using namespace at;
+
+// Some operations can be performed more efficiently if we're handling tensors
+// of a single type only. Adding this logic directly in the loop makes it a bit
+// ugly, so here's a helper for it.
+struct unique_type_checker {
+  void show(const at::Type& t) {
+    if (!unique) return;
+    if (!type) type = &t;
+    unique = (type == &t);
+  }
+
+  const at::Type *type = nullptr;
+  bool unique = true;
+};
+
+std::vector<Tensor> broadcast(const Tensor& tensor, const IntList& devices) {
+  auto & type = tensor.type();
+  if (type.is_cuda() && tensor.get_device() != devices[0])
+    throw std::runtime_error("device of broadcasted tensor must appear as the "
+                             "first on devices list");
+  std::vector<Tensor> tensors;
+  tensors.reserve(devices.size());
+  if (nccl::is_available({tensor})) {
+    tensors.push_back(tensor);
+    for (std::size_t i = 1, num_devices = devices.size(); i < num_devices; ++i) {
```

Comment:
Can this just be:

```
for (int device : devices) {
 ...
}
```

Reply:
It's from 1 - first element is just the original tensor

Reply:
FWIW, you could do:

```
for (int device : devices.slice(1)) {
}
```

But it doesn't matter.

Reply:
Oh nice. I'll use that